File: styled_text.cpp
#include "styled_text.h"
#include "formatted.h"
#include <boost/algorithm/string/join.hpp>

/************************************************************************
 * Static data
 ***********************************************************************/

vector<pair<styled_text::color_e, U16>> styled_text::color_data {
    { styled_text::color_none, 0 },
    { styled_text::black, 232 },
    { styled_text::red, 9 },
    { styled_text::green, 40 },
    { styled_text::yellow, 11 },
    { styled_text::blue, 20 },
    { styled_text::magenta, 90 },
    { styled_text::cyan, 14 },
    { styled_text::white, 15 },
    { styled_text::grey, 244 },
    { styled_text::deep_blue, 20 },
    { styled_text::mid_blue, 27 },
    { styled_text::orange, 208 },
    { styled_text::pink, 201 },
    { styled_text::brown, 1 },
};

vector<pair<styled_text::style_e, U16>> styled_text::style_data {
    { styled_text::style_none, 0 },
    { styled_text::bold, 1 },
    { styled_text::italic, 3 },
    { styled_text::blink, 5 },
    { styled_text::underline, 4 },
    { styled_text::crossed, 9 },
    { styled_text::inverted, 7 },
};

vector<U16> styled_text::color_map;
vector<U16> styled_text::style_map;

string styled_text::escape_ch = "\u001b";

styled_text::render_e styled_text::renderer = plain;

/************************************************************************
 * Constructors
 ***********************************************************************/

styled_text::styled_text(const string &t, color_e c, color_e bg, style_e s, size_t sz)
    : text(t), color(c), background(bg), style(s), size(sz)
{
}

styled_text::styled_text(const styled_text &st)
    : text(st.text), color(st.color), background(st.background), style(st.style), size(st.size)
{
    for (const styled_text &t : st.subtexts) {
        subtexts.emplace_back(t);
    }
}

styled_text &styled_text::append(const string &t)
{
    if (subtexts.empty()) {
        text += t;
    } else {
        subtexts.back().text += t;
    }
    return *this;
}

styled_text &styled_text::append(const styled_text &st)
{
    if (subtexts.empty() && !text.empty()) {
        subtexts.emplace_back(*this);
        text.clear();
    }
    subtexts.emplace_back(st);
    return *this;
}


styled_text &styled_text::override(color_e c, color_e bg, style_e s, size_t sz)
{
    set_color(c);
    set_background(bg);
    set_style(s);
    set_size(sz);
    return *this;
}

styled_text &styled_text::override(const styled_text &st)
{
    set_color(st.color);
    set_background(st.background);
    set_style(st.style);
    set_size(st.size);
    return *this;
}

styled_text &styled_text::underride(color_e c, color_e bg, style_e s, size_t sz)
{
    underride_color(c);
    underride_background(bg);
    underride_style(s);
    underride_size(sz);
    return *this;
}


styled_text &styled_text::underride(const styled_text &st)
{
    underride_color(st.color);
    underride_background(st.background);
    underride_style(st.style);
    underride_size(st.size);
    return *this;
}

string styled_text::str() const
{
    if (subtexts.empty()) {
        return text;
    } else {
        vector<string> strs;
        for (const styled_text &st : subtexts) {
            strs.emplace_back(st.str());
        }
        return boost::algorithm::join(strs, "");
    }
}

string styled_text::render(render_e r) const
{
    switch (r) {
    case plain:
        return render_plain();
    case iso6429:
        return render_iso6429();
    default:
        return render_plain();
    }
}

string styled_text::render_plain() const
{
    return str();
}

string styled_text::render_iso6429() const
{
    make_maps();
    string result;
    if (subtexts.empty()) {
        result = render_style_iso6429(style) +
            render_color_iso6429(color, false) +
            render_color_iso6429(background, true) +
            text;
    } else {
        for (const styled_text &st : subtexts) {
            styled_text st1(st);
            st1.underride(*this);
            result += st1.render_iso6429();
        }
    }
    result += render_style_iso6429(style_none);
    return result;
}

string styled_text::render_color_iso6429(color_e c, bool bg)
{
    U16 op = bg ? op_bg : op_fg;
    string result;
    if (c==color_none) {
        result = formatted("\e[%dm", op + 1); 
    } else {
        result = formatted("\e[%d:5:%dm", op, color_map[c]);
    }
    return result;
}

string styled_text::render_style_iso6429(style_e s)
{
    string result;
    if (s == style_none) {
        result = "\e[0m";
    } else {
        U16 ss = s;
        while (ss != 0) {
            U16 ss_new = ss & (ss - 1);
            U16 this_s = ss ^ ss_new;
            ss = ss_new;
            result += formatted("\u001b[%dm", style_map[this_s]);
        }
    }
    return result;
}

void styled_text::make_maps()
{
    if (color_map.empty()) {
        color_map.resize(color_final);
        for (const auto &cd : color_data) {
            color_map[cd.first] = cd.second;
        }
        style_map.resize(style_final);
        for (const auto &sd : style_data) {
            style_map[sd.first] = sd.second;
        }
    }
}

File: styled_text.py
from __future__ import annotations
from enum import Enum
from typing import Iterable
import sys
import os

class StyledText:

    render_style_plain = 0
    render_style_iso6429 = 1

    render_style = 1 if sys.stdout.isatty() else 0

    color_map = {
        'none' : 0,
        'black' : 232,
        'red' : 9,
        'green' : 40,
        'dark_green' : 34,
        'yellow' : 11,
        'blue' : 20,
        'magenta' : 90,
        'cyan' : 14,
        'white' : 15,
        'grey' : 244,
        'deep_blue' : 20,
        'mid_blue' : 27,
        'orange' : 208,
        'pink' : 201,
        'brown' : 1,
    }

    style_map = {
        'none' : 0,
        'bold' : 1,
        'italic' : 3,
        'blink' : 5,
        'underline' : 4,
        'crossed' : 9,
        'inverted' : 7,
    }

    def __init__(self,
                 text: str|StyledText|Iterable[StyledText] = '',
                 color: str|None = None,
                 background: str|None = None,
                 style: str|Iterable[str]|None = None,
                 size: float=0 ):
        self.color = color
        self.background = background
        self.style = style
        self.size = size
        self.subtexts: list[StyledText] = []
        if isinstance(text, str):
            self.text = text
        elif isinstance(text, StyledText):
            self.text = ''
            self.subtexts = [ text ]
        else:
            self.text = ''
            self.subtexts = [ s for s in text ]

    def copy(self) -> StyledText:
        st = StyledText(self.text,
                        color=self.color,
                        background=self.background,
                        style = self.style,
                        size = self.size)
        st.subtexts = self.subtexts.copy()
        return st

    def _make_subtext(self) -> StyledText:
        if not self.subtexts and self.text:
            self.subtexts = [ self.copy() ]
        self.text = ''
        return self

    def __iadd__(self, other: str|StyledText) -> StyledText:
        self._make_subtext()
        if isinstance(other, str):
            self.subtexts.append(StyledText(other))
        else:
            self.subtexts.append(other)
        return self

    def __add__(self, other: str|StyledText) -> StyledText:
        result = self.copy()
        result += other
        return result

    def attach(self,
               text: str|StyledText,
               color: str|None = None,
               background: str|None = None,
               style: str|Iterable[str]|None = None,
               size: float=0 ) -> StyledText:
        self._make_subtext()
        if isinstance(text, str):
            self.subtexts.append(StyledText(text, color=color, background=background, style=style, size=size))
        else:
            self.subtexts.append(text.copy())
        self.subtexts[-1].underride_from(self)
        return self

    def override(self,
                 color: str|None = None,
                 background: str|None = None,
                 style: str|Iterable[str]|None = None,
                 size: float=0 ) -> StyledText:
        self.color = color or self.color
        self.background = background or self.background
        self.style = style or self.style
        self.size = size or self.size
        return self

    def underride(self,
                 color: str|None = None,
                  background: str|None = None,
                 style: str|Iterable[str]|None = None,
                 size: float=0 ) -> StyledText:
        self.color = self.color or color
        self.background = self.background or background
        self.style = self.style or style
        self.size = self.size or size
        return self

    def override_from(self, other: StyledText) -> StyledText:
        self.override(color=other.color, background=other.background, style=other.style, size=other.size)
        return self

    def underride_from(self, other: StyledText) -> StyledText:
        self.underride(color=other.color, background=other.background, style=other.style, size=other.size)
        return self

    def __str__(self) -> str:
        return self.render()

    def get_text(self) -> str:
        if self.subtexts:
            return ''.join([ s.get_text() for s in self.subtexts ])
        else:
            return self.text

    def render(self) -> str:
        match StyledText.render_style:
            case 0:
                return self.get_text()
            case 1:
                return self.render_iso6429()
            case _:
                return self.get_text()

    def render_iso6429(self) -> str:
        esc = '\u001b'
        csi = esc + '['
        result = ''
        if self.subtexts:
            result = ''.join([ s.render_iso6429() for s in self.subtexts ])
        else:
            if self.style:
                for ss in [self.style] if isinstance(self.style, str) else self.style:
                    result += f'{csi}{self.get_style(ss)}m'
            else:
                result += f'{csi}0m'
            if self.color:
                result += f'{csi}38;5;{self.get_color(self.color)}m'
            else:
                result += f'{csi}39m'
            if self.background:
                result += f'{csi}48;5;{self.get_color(self.background)}m'
            else:
                result += f'{csi}49m'
            result += self.text
        result += f'{csi}39m{csi}49m{csi}0m'
        #print(' '.join([ str(ord(b)) for b in result]))
        #print(' '.join([ f"'{b}'" if b!=esc else 'ESC' for b in result]))
        return result

    def get_color(self, c: str) -> int:
        return self._get_from_map(StyledText.color_map, c, 'none')

    def get_style(self, s:str) -> int:
        return self._get_from_map(StyledText.style_map, s, 'none')

    def _get_from_map(self, map: dict[str, int], s: str, dflt: str) -> int:
        try:
            return map[s]
        except KeyError:
            try:
                return int(s)
            except ValueError:
                return map[dflt]

    @staticmethod
    def set_render_style_plain() -> None:
        StyledText.render_style = 0

    @staticmethod
    def set_render_style_iso6429() -> None:
        StyledText.render_style = 1

if __name__=='__main__':
    print(StyledText('normal'))
    print(StyledText('red', color='red'))
    print(StyledText('green underline', color='green', style='underline'))
    print(StyledText('blue italic underline', color='blue', style=('italic', 'underline')))
    s1 = StyledText('magenta underline', color='magenta', style='underline')
    s1.attach(' orange underline', color='orange')
    print(s1.get_text(), s1)
    s3 = s1.attach(' orange italic', color='orange', style='italic')
    print(s3)
    s4 = StyledText((StyledText('pink', color='pink'), StyledText('bold', style='bold')))
    print(s4)
    s5 = StyledText((StyledText('pink', color='pink'),
                     StyledText((StyledText('bold', style='bold'),
                                StyledText('brown inverted', color='brown', style='inverted')))))
    print(s5)
    print('as text: ', s5.get_text())
    s6 = StyledText('cyan', color='pink')
    s6.override(color='cyan')
    print(s6)
    s7 = StyledText('magenta italic', color='magenta', style='italic')
    s7.underride(color='orange', style='bold')
    print(s7)
    print(StyledText('orange on grey', color='orange', background='grey'))
    s8 = StyledText('green', color='green').attach('blue', color='blue').attach('green')
    print(s8)
    s9 = StyledText('pink bold', color='pink', style='bold') + StyledText('normal')
    print(s9)
    print(StyledText('normal'))
    
        
                
